logstash-core/lib/logstash/util/safe_uri.rb

```diff
@@ -37,20 +38,132 @@ def inspect
   end
 
   def sanitized
-    return uri unless uri.password # nothing to sanitize here!
+    return uri unless password # nothing to sanitize here!
     
-    safe = uri.clone
-    safe.password = PASS_PLACEHOLDER
-    safe
+    user_info = user ? "#{user}:#{PASS_PLACEHOLDER}" : nil
+
+    # This syntax is required for jruby 1.7.x which doesn't reflect here correctly
+    # 9k does reflect here correctly. This can be removed when we are fully jruby 9k 
+    constructor = java.net.URI.java_class.constructor(java.lang.String, java.lang.String, java.lang.String, Java::int, java.lang.String, java.lang.String, java.lang.String)
+    java.net.URI.new(scheme, user_info, host, port || -1, path, query, fragment)
   end
 
   def ==(other)
     other.is_a?(::LogStash::Util::SafeURI) ? @uri == other.uri : false
   end
 
   def clone
-    cloned_uri = uri.clone
-    self.class.new(cloned_uri)
+    # No need to clone the URI, in java its immutable
+    self.class.new(uri)
+  end
+
+  def update(field, value)
+    new_scheme = scheme
+    new_user = user
+    new_password = password
+    new_host = host
+    new_port = port
+    new_path = path
+    new_query = query
+    new_fragment = fragment
+
+    case field 
+    when :scheme
+      new_scheme = field
+    when :user
+      new_user = field
+    when :password
+      new_password = field
+    when :host
+      new_host = field
+    when :port
+      new_port = field
+    when :path
+      new_path = field
+    when :query
+      new_query = field
+    when :fragment
+      new_fragment = field
+    end
+
+    user_info = new_user
+    if new_user && new_password
+      user_info += ":" + new_password
```

Comment:
Thinking through this a bit more, I think its not a good idea to escape the user/password.

Both the java.net.URI and ruby's URI class don't do this and I think it makes sense not to include this feature.

I imagine using a URI to represent, say, a rabbitmq connection. When using the rabbitmq libs you wouldn't want to do the encoding, that library would do it for you, as do some HTTP libraries. If you need the raw password in that case you have to unescape it or use a special method.